[lldb/formatter] Add Swift.Unsafe[Mutable][Raw]Pointer data formatter

medismailben opened the pull request:
This patch adds support for Swift's Unsafe[Mutable]Pointer and
UnsafeRaw[Mutable]Pointer data formatting.

Since the 2 types have a similar layout, this patch makes use of the
SwiftUnsafeType factory to create a single helper class for the 2 types.

This commit also removes the old basic summary provider to these 2 types.

rdar://64410428

Signed-off-by: Med Ismail Bennani <medismail.bennani@gmail.com>

medismailben commented:
@swift-ci test

medismailben commented:
@swift-ci test

medismailben commented:
@swift-ci test platform Linux

medismailben commented:
@swift-ci test

medismailben commented:
This PR solves the 2 failing test in #1349 and moves the patch from `swift/master` to `swift/master-rebranch`:

```
Failing Tests (2):
    lldb-api :: functionalities/asan/swift/TestAsanSwift.py
    lldb-api :: functionalities/swift-runtime-reporting/exclusivity-violation/TestExclusivityViolation.py
```

The code review has already been done in #1349.

The linux bot already had the `clangd` build failure on `swift/master-rebranch`:
https://ci.swift.org/view/swift-master-rebranch/job/oss-swift-master-rebranch-lldb-incremental-linux-ubuntu-16_04/

On the macOS bots, these tests have been failing prior to this PR.
```
Failed Tests (2):
  lldb-api :: lang/swift/tagged_pointer/TestSwiftTaggedPointer.py
  lldb-api :: lang/swift/variables/dict_nsobj_anyobj/TestSwiftDictionaryNSObjectAnyObject.py
```

I think it's safe to merge this PR.

adrian-prantl commented on lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp:
```diff
@@ -263,6 +343,70 @@ bool SwiftUnsafeRawBufferPointer::Update() {
+    return false;
+  }
+
+  SwiftASTContext *type_system = GetSwiftASTContextTypeSystem(type);
```
Instead of adding `GetSwiftASTContextTypeSystem()` here, it would be better to hoist `GetGenericArgumentType()` into the `TypeSystemSwift` base class and provide a default implementation using ReconstructType to `TypeSystemSwiftTypeRef`  (Or, even better, proivide a real implementation). Then you can just cast `type.GetTypeSystem(`) to `TypeSystemSwift` here and use that.

adrian-prantl commented on lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp:
```diff
@@ -64,11 +75,12 @@ lldb::addr_t SwiftUnsafeType::GetAddress(llvm::StringRef child_name) {
   }
 
-  auto type_system = llvm::dyn_cast<SwiftASTContext>(type.GetTypeSystem());
+  SwiftASTContext *type_system = GetSwiftASTContextTypeSystem(type);
```
Now this should just be 

`auto *typesystem = dyn_cast_or_null<TypeSystemSwift>(type.GetTypeSystem())`

and GetSwiftASTContextTypeSystem can be removed, right?

medismailben replied:
Indeed!

medismailben committed:
[lldb/formatter] Add Swift.Unsafe[Mutable][Raw]Pointer data formatter

This patch adds support for Swift's Unsafe[Mutable]Pointer and
UnsafeRaw[Mutable]Pointer data formatting.

Since the 2 types have a similar layout, this patch makes use of the
SwiftUnsafeType factory to create a single helper class for the 2 types.

This commit also removes the old basic summary provider to these 2 types.

rdar://64410428

Signed-off-by: Med Ismail Bennani <medismail.bennani@gmail.com>

adrian-prantl commented on lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp:
```diff
@@ -64,15 +74,17 @@ lldb::addr_t SwiftUnsafeType::GetAddress(llvm::StringRef child_name) {
+      llvm::dyn_cast_or_null<TypeSystemSwift>(type.GetTypeSystem());
   if (!type_system) {
     LLDB_LOG(GetLogIfAllCategoriesSet(LIBLLDB_LOG_DATAFORMATTERS),
              "{0}: Couldn't get {1} type system.", __FUNCTION__,
```
I think "Not a Swift type" would be a more accurate log message.

adrian-prantl commented on lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp:
```diff
@@ -263,6 +277,71 @@ bool SwiftUnsafeRawBufferPointer::Update() {
+      llvm::dyn_cast_or_null<TypeSystemSwift>(type.GetTypeSystem());
+  if (!type_system) {
+    LLDB_LOG(GetLogIfAllCategoriesSet(LIBLLDB_LOG_DATAFORMATTERS),
+             "{0}: Couldn't get {1} type system.", __FUNCTION__,
```
same here

adrian-prantl commented on lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp:
```diff
@@ -263,6 +277,71 @@ bool SwiftUnsafeRawBufferPointer::Update() {
+    return false;
+  }
+
+  addr_t addr = pointer_value_sp->GetValueAsUnsigned(LLDB_INVALID_ADDRESS);
```
We should really convert GetValueAsUnsigned to return an optional at one point...

adrian-prantl commented on lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp:
```diff
@@ -924,6 +924,13 @@ CompilerType TypeSystemSwiftTypeRef::GetTypeFromMangledTypename(
+TypeSystemSwiftTypeRef::GetGenericArgumentType(opaque_compiler_type_t type,
+                                               size_t idx) {
+  return m_swift_ast_context->GetGenericArgumentType(ReconstructType(type),
+                                                     idx);
```
Can I interest you in writing a real implementation for this at one point int the future?

adrian-prantl commented:
Please don't forget to cherry-pick to master-next, too!
